### ISSUE TITLE
Following upgrade to v24.02.1-93-gc4edec8 can't start service due to missing module 'pyln'

### ISSUE THREAD

**Issue by unsenttraffic** [ElementsProject/lightning#7220]
I'm running a node on a Pi 4B.

The output of 'cat /etc/os-release' is:
```
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
```

I originally installed Core Lightning following the RaspiBolt documentation (https://raspibolt.org/guide/bonus/lightning/cln.html) with reference to the Elements documentation here. I was on v23.08 and am attempting to upgrade to v24.02.1-93-gc4edec8.

Although the make process is complete and 'lightning-cli --version' returns v24.02.1-93-gc4edec8, it's not possible to start the service due to a missing module:
```
lightningd: v24.02.1-93-gc4edec8
plugin-clnrest: Killing plugin: disabled itself: No module named 'pyln'
```

Prior to this missing module, each attempt to run 'systemctl restart lightningd.service' would fail due to other missing modules. I was able to install the following:
grpcio-tools, gevent, gunicorn, flask, flask_restx, flask_cors, flask_socketio, json5

However, the output of 'python3 -m pip install --user pyln' is:
```
ERROR: Could not find a version that satisfies the requirement pyln
ERROR: No matching distribution found for pyln
```

I can't locate any package named 'pyln' under https://pypi.org/search/?q=pyln

I initially created a thread on bitcointalk logging my attempts to resolve the issue (https://bitcointalk.org/index.php?topic=5492706.0), but I hope all relevant information is included above.

**Comment by vincenzopalazzo** [comment]
if you are not using `plugin-clnrest` plugin this message can be ignored because it is just the plugin that is not able to start. clnrest is not a very important plugin if you are not using the Rest API.

**Comment by unsenttraffic** [comment]
> if you are not using `plugin-clnrest` plugin this message can be ignored because it is just the plugin that is not able to start. clnrest is not a very important plugin if you are not using the Rest API.

In that case, I need to ensure Core Lightning ignores the message and starts regardless. Would that mean simply running `reckless uninstall plugin-clnrest`?

I assume that wouldn't affect use of lightning-cli to manage the node?

**Comment by vincenzopalazzo** [comment]
>Would that mean simply running reckless uninstall plugin-clnrest?

idk I am not a reckless users, idk how it works, but I on my cloud node I use `--disable-plugin=clnrest`

**Comment by unsenttraffic** [comment]
I tried adding 'disable-plugin=clnrest' and later 'disable-plugin=clnrest.py' (as per https://docs.corelightning.org/docs/rest#installation) to the config file and starting, but this made no difference.

Trying the syntax 'sudo systemctl start lightningd.service --disable-plugin=clnrest' or 'sudo systemctl start lightningd.service --disable-plugin=clnrest.py' returns 'systemctl: unrecognized option '--disable-plugin=clnrest'' or 'systemctl: unrecognized option '--disable-plugin=clnrest.py'.'

**Comment by unsenttraffic** [comment]
Using journalctl, I found that one reason for the exit with failure is given as:
`lightningd: Config file /data/lightningd/config line 18: autocleaninvoice-cycle=86400: unknown option`

I don't think it wise to set it to 0, but when I did the reason for the exit with failure changed to:
`Refusing to irreversibly upgrade db from version 233 to 256 in non-final version v24.02.1-93-gc4edec8-modded (use --database-upgrade=true to override)

I don't think it wise to do this, but I'd be grateful for any advice.

**Comment by vincenzopalazzo** [comment]
Ah you need to add the `--database-upgrade=true`

**Comment by unsenttraffic** [comment]
> Ah you need to add the `--database-upgrade=true`

I understand I have the option of adding that to force an override, my question is whether or not that's a good idea. The refusal to upgrade the db must stem from the fact that the version changed to -modded after I installed the missing requirements pyln-bolt7, pyln-client and pyln-proto and went through the ./configure and make steps again. 

But two serious questions:
- why am I not able to use an official released version?
- why am I required to remove autocleaninvoice and autocleaninvoice-expired-by config settings?

**Comment by vincenzopalazzo** [comment]
>I understand I have the option of adding that to force an override, my question is whether or not that's a good idea. 

the upgrade means, if you are running a not release version like `v24.04` you need to tell lightning that he can upgrade the database, because if the database will upgrade you can not reverse this operation.

So the question is why you are running  `v24.02.1-93-gc4edec8` a not tagged version?

**Comment by unsenttraffic** [comment]
> So the question is why you are running `v24.02.1-93-gc4edec8` a not tagged version?

Answered above - due to the manual installation of required modules, in turn due to a failed upgrade process.

The outstanding questions are:
- why am I not able to use an official released version?
- why am I required to remove autocleaninvoice and autocleaninvoice-expired-by config settings?
- does this mean that the db has not been changed and that a downgrade would be a possible solution?

**Comment by vincenzopalazzo** [comment]
>why am I not able to use an official released version?

probably a bug

>why am I required to remove autocleaninvoice and autocleaninvoice-expired-by config settings?

you should not, maybe it just a log aggregation between runs

>does this mean that the db has not been changed and that a downgrade would be a possible solution?

yes

**Comment by unsenttraffic** [comment]
> probably a bug
That's why I've opened this issue.
> you should not, maybe it just a log aggregation between runs
Then why did Core Lightning exit with failure, returning:
`lightningd: Config file /data/lightningd/config line 18: autocleaninvoice-cycle=86400: unknown option`
> yes
That implies that the software can't be upgraded, i.e. that it's unusable. I'd rather not jump to that conclusion as hastily as you have.

**Comment by vincenzopalazzo** [comment]
> That's why I've opened this issue.

The issue is describing another kind of problem, not the dirty build IMHO

>That implies that the software can't be upgraded, i.e. that it's unusable. I'd rather not jump to that conclusion as hastily as you have.

wow I am really confused, if you want use it with a dirty version like in your case, you should specify `--database-upgrade=true`, if you are building from a tagged version you are safe to do so.

**Comment by unsenttraffic** [comment]
I give up.

**Comment by vincenzopalazzo** [comment]
Sorry about that, let make this simple:

Are you compiling from a tagged version like `v24.04`?
- yes: add the `--database-upgrade=true` because it is a bug of our build system
- no: check out the tagged version that you want to build and add the `--database-upgrade=true`

Report what you get and we move forward to the next error

**Comment by unsenttraffic** [comment]
This is basically resolved at my end now, by luck rather than skill. For what it's worth, I'll share what I did as briefly as possible:

- Although git pull would leave me with a -modded version, git pull --rebase brought it back to the official release version. I thought it best to keep the new files so I added --autostash.
- The service still exited with failure after specifying --database-upgrade=true. The logs showed a plugin file not being recognised.
- I then tried to go backwards. I tried a hard git reset back to v23.08, which compiled successfully then exited with failure refusing to reverse database changes.
- I was then able to use git config pull.rebase true to move forward to v24.02.1-93-gc4edec8. Without specifying database-upgrade it started successfully and synced blocks. I do still have autoclean options commented out in my config, I'll try to restore them on my next upgrade.
